core/src/main/java/org/kiwix/kiwixmobile/core/downloader/ChunkUtils.kt

```diff
@@ -0,0 +1,101 @@
+/*
+ * Kiwix Android
+ * Copyright (c) 2019 Kiwix <android.kiwix.org>
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+package org.kiwix.kiwixmobile.core.downloader
+
+import org.kiwix.kiwixmobile.core.utils.StorageUtils.getFileNameFromUrl
+import java.util.ArrayList
+import java.util.Locale
+
+object ChunkUtils {
+  private const val ALPHABET = "abcdefghijklmnopqrstuvwxyz"
+  private const val ZIM_EXTENSION = ".zim"
+  const val PART = ".part.part"
+  const val CHUNK_SIZE = 1024L * 1024L * 1024L * 2L
+
+  fun getChunks(
+    url: String?,
+    contentLength: Long,
+    notificationID: Int
+  ): List<Chunk> {
+    val fileCount = getZimChunkFileCount(contentLength)
+    val filename = getFileNameFromUrl(url)
+    val fileNames = getZimChunkFileNames(filename, fileCount)
+    return generateChunks(contentLength, url, fileNames, notificationID)
+  }
+
+  private fun generateChunks(
+    contentLength: Long,
+    url: String?,
+    fileNames: Array<String?>,
+    notificationID: Int
+  ): List<Chunk> {
+    val chunks: MutableList<Chunk> = ArrayList()
+    var currentRange: Long = 0
+    for (zim in fileNames) {
```

Comment:
there is definitely a nicer way of expressing this functionally. Lets iterate on this a bit.

For starters lets reduce some noise

Use kotlin Ranges to express this `currentRange + CHUNK_SIZE >= contentLength`

create a ChunkFactory that takes the data that doesn't change from chunk to chunk.

Move ` currentRange += CHUNK_SIZE + 1` out of the if and the else because it is common to both, just put it at the bottom of the loop.

`var range: String` has too broad a scope, inline it or turn it into 2 vals.

Then maybe some combination of zip/map/fold will look good

Reply:
@macgills I have made a few changes. Please review it.

Reply:
@sonusourav I would still like to see this

>Use kotlin Ranges to express this currentRange + CHUNK_SIZE >= contentLength

You can read more on ranges [here](https://kotlinlang.org/docs/reference/ranges.html)

I am also trying to steer you towards this

> Then maybe some combination of zip/map/fold will look good

and trying to push the idea that any for loop can be written using higher order functions instead.

It would help for you to identify what the inputs are to each iteration but depending on your familiarity with these concepts it could be quite tough. If you want to press ahead with learning about these things I am happy to assist but if you are fine right now then we can merge this PR before we get to `zip/map/fold` portion of this comment

Reply:
Hey give me one more chance, I will try to implement it with your help. I appreciate that you care we should learn new things and not simply be worried about getting PRs merged 😃

Reply:
Hey, I did the range one but I am not getting the use case of zip/map/fold or for loop. Could you please provide some hint

Reply:
The end goal is you want a simple operation to create a chunk. if chunk was a simpler piece of data we could imagine this as
```
val chunks = zimFiles.map { Chunk(it) }
return chunks
```
which could of course be rewritten to not delcare a variable at all (which is what I want to see)
```
return zimFiles.map { Chunk(it) }
```

but we need more input information than just the zimfile to create a chunk.

`currentRange` should no longer be a long but a kotlin range.

Reply:
@sonusourav it has been 8 days since the last commit, will you have time to finish this PR or do you need more info about what I want to see?

Reply:
Hey @macgills sorry, I had few course commitments to do. I have wrapped them now. 
Actually, I new some more hints on how to incorporate the logic of minLen and range in map. 
At some point, I got a feeling that something like an accumulator and current value in higher function could help, but I am not getting the exact point. 
Please guide me on this.

Reply:
So the idea I was trying to get across was that the input to each iteration is
1. a zimfile
1. a range

and the output of each iteration is a chunk.


to express that a zip is what we could use

```
val filenamesAndRanges:List<Pair<String, Range>> =filenames.zip(ranges)
```

we can use lambda parameter destructuring to make the next step easier

```
filenamesAndRanges.map{ (filename, currentRange) -> createAChunkWith(filename, range) }
```

turns out the fold was unnecessary, sorry for the misdirect, while it is possible to write it with a fold it is not necessary

Reply:
@macgills  Please review this one.